### PR TITLE
[inductor] Rewind block_ptr across outer loop in nested reductions (#188771)

### PR DESCRIPTION
**Pull request by limintang**
Summary:

For a fused broadcast + reduction (e.g. `mean(a * b)` where `a:[N]` and `b:[N, 1]` broadcast to `[N, N]`), Inductor's Triton codegen emits nested `for` loops over the two reduction axes. Block pointers are threaded through the loops and advanced with `tl.advance` at the suffix of each level.

The bookkeeping in `TritonKernel` recorded a pointer's per-level advancement only when that advancement was non-zero (`triton.py` around L3999-4005 skipped identity offsets). For a pointer that is constant along the outer axis but stepped along the inner axis, this meant no entry was recorded for the outer axis, so no `tl.advance` was emitted at the outer-loop suffix. On backends that lower the Python `for` to `scf.for` with the block pointer as an iter-arg (e.g. Triton-MTIA), the SSA value yielded at the end of the outer iteration was the pointer already stepped through all inner iterations. Subsequent outer iterations would then load out-of-bounds, silently returning padding (zero) and dropping most of the reduction. For a `6144x6144` case with `R0_BLOCK=R1_BLOCK=32` the observed result was `1/192` of the true value.

**The bug is generic, not MTIA-specific.** Triton lowers every backend's Python for-loop to `scf.for`, and any variable reassigned inside the loop (like `block_ptr = tl.advance(block_ptr, ...)`) becomes an iter-arg carried across iterations. CUDA Triton is mechanically susceptible to the same OOB-load pattern. That the bug hasn't surfaced on CUDA before is likely a combination of (1) MTIA's autotuner floors `R0_BLOCK` at 32 whereas CUDA typically picks a much larger tile, so on CUDA the "iteration 0 only" truncation drops a much smaller fraction of the reduction and is easy to attribute to BF16 accumulation noise; (2) Inductor's scheduler more often avoids nested-reduction codepaths on CUDA for broadcast-heavy shapes; and (3) most CUDA tests use zero-mean inputs where a truncated outer-product reduction is close to zero anyway. Landing this as an upstream Inductor fix rather than an MTIA-only workaround.

The fix records every pointer at every reduction level, including zero offsets, so the existing outer-suffix subtraction `outer_step - inner_step * inner_num_iter` can compute the correct rewind (e.g. `0 - 32*192 = -6144`). To keep the emitted kernel body clean, the emit site now drops `tl.advance(ptr, [0, ...])` lines whose final advancement collapses to zero. Backends that keep the block pointer as a Python-local rebound each iteration are unaffected; backends that thread it through SSA iter-args now see the rewind they need.

Test Plan:
Added a regression test `test_nested_reduction_broadcast_outer_product` in `caffe2/test/inductor/test_torchinductor.py` that exercises `(a * b).sum()` and `torch.mean(a * b)` on `[N]` and `[N, 1]` inputs at `N=1024`, using non-zero-mean inputs so a truncated reduction cannot hide behind BF16 noise.

```
$ buck2 test mode/opt //caffe2/test/inductor:test_inductor -- test_nested_reduction_broadcast_outer_product
```

Also verified out-of-tree on MTIA-Triton with a shifted-mean bcast harness (`randn + 3.0` at `N=6144`):

| Case | Pre-fix `mtia_err / cpu_bf16_err` | Post-fix ratio |
| --- | --- | --- |
| `sum([N] * [N,1])` | 689x wrong | 1.0 (matches) |
| `mean([N] * [N,1], shift=10)` | 436x wrong | 1.0 (matches) |
| `mean([N,22] * [N,1])` (production LAMP shape) | 1.0 (unaffected) | 1.0 (unaffected) |

Confirmed the generated Triton IR after the fix contains the missing outer-loop rewind: `tt.advance %8#1, [-6144]`.

Authored with AI assistant (Claude Code).

Differential Revision: D110408213

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo

### CODE REVIEW

**Comment by jansel** [comment]
@claude explain those CI failures, are they related to this PR?

**Comment by limintang** [comment]
@jansel Could you review again? The remaining test failures seem to be infra issue, unrelated to the PR.

**Review by jansel — APPROVED** [review]
